op-chain-ops: add check-jovian cmd

geoknee opened the pull request:
```
➜  check-jovian git:(gk/check-jovian) ✗ CHECK_JOVIAN_L2=<jovian-l2-rpc> CHECK_JOVIAN_SECRET_KEY=<funded-key> go run main.go all
INFO [11-20|13:37:10.438] starting Jovian checks
INFO [11-20|13:37:10.831] GasPriceOracle test: success             isJovian=true
INFO [11-20|13:37:10.959] L1Block test: success                    daFootprintGasScalar=400
INFO [11-20|13:37:10.959] secret key provided - attempting to send tx-to-self and wait for inclusion
INFO [11-20|13:37:11.712] Publishing transaction                   component=txmgr service=check-jovian tx=0x7fe6f38aef4e46065b32397b71dc161a643a9ec1fe2a9ff9fefee613f0505655 nonce=2160 gasTipCap=1,000,000,000 gasFeeCap=3,000,000,000 gasLimit=21000
INFO [11-20|13:37:11.835] Transaction successfully published       component=txmgr service=check-jovian tx=0x7fe6f38aef4e46065b32397b71dc161a643a9ec1fe2a9ff9fefee613f0505655 nonce=2160 gasTipCap=1,000,000,000 gasFeeCap=3,000,000,000 gasLimit=21000 tx=7fe6f3..505655
INFO [11-20|13:37:23.836] Publishing transaction                   component=txmgr service=check-jovian tx=0x7fe6f38aef4e46065b32397b71dc161a643a9ec1fe2a9ff9fefee613f0505655 nonce=2160 gasTipCap=1,000,000,000 gasFeeCap=3,000,000,000 gasLimit=21000
INFO [11-20|13:37:36.213] Transaction confirmed                    component=txmgr service=check-jovian tx=7fe6f3..505655 block=21a8ae..e554fd:1153386 effectiveGasPrice=1,000,000,251
WARN [11-20|13:37:36.214] Unable to fetch blob base fee            component=txmgr service=check-jovian err="Post \"https://jovian-alpha-opn-geth-f-sequencer-0-op-geth.primary.infra.dev.oplabs.cloud/\": context canceled"
INFO [11-20|13:37:36.214] Transaction confirmed                    component=txmgr service=check-jovian tx=7fe6f3..505655 block=21a8ae..e554fd:1153386 effectiveGasPrice=1,000,000,251
INFO [11-20|13:37:36.214] tx mined                                 txHash=0x7fe6f38aef4e46065b32397b71dc161a643a9ec1fe2a9ff9fefee613f0505655 blockNumber=1,153,386 blobGasUsed=40000
INFO [11-20|13:37:36.338] Block header test: success - non-zero BlobGasUsed is hard evidence of Jovian being active blockNumber=0x102962570 blobGasUsed=40000 expectedDAFootprint=40000
INFO [11-20|13:37:36.463] ExtraData format test: success           blockNumber=1,153,398   version=1 denominator=250 elasticity=6 minBaseFee=0
INFO [11-20|13:37:36.463] completed all tests successfully!
```
➜  optimism git:(develop) ✗ CHECK_JOVIAN_L2="https://mainnet.optimism.io" go run op-chain-ops/cmd/check-jovian/main.go all
INFO [11-13|12:09:18.236] starting Jovian checks
Application failed: command error: failed: GPO contract error: failed to get jovian status: execution reverted
exit status 1
```

geoknee committed:
op-chain-ops: add check-jovian cmd

geoknee commented on op-chain-ops/cmd/check-jovian/main.go:
```diff
@@ -0,0 +1,169 @@
+	}
+	if latest.BlobGasUsed == nil {
+		return fmt.Errorf("block %d has nil BlobGasUsed field", latest.Number)
+	}
```
This doesn't really show that Jovian is active: to do that, there must be some traffic (user transactions) on the chain, and then the value would be nonzero. This can be tested manually (our acceptance tests use a tx spammer to do this).

sebastianst replied:
Yeah, but as a first smoke test that's already good, because only with Jovian is that field non-nil.

If a more comprehensive correctness test is needed, then the acceptance tests should be ran.

geoknee replied:
Actually, it is zero pre-Jovian (not nil) https://specs.optimism.io/protocol/jovian/exec-engine.html#da-footprint-block-limit. This step passes on OPM (pre Jovian).

sebastianst replied:
Ohh, good callout, we only omitted it in receipts where it is added with Jovian. Then this check is indeed invalid/useless. Let's print a message like "Zero blobGasUsed is inconclusive" and say that the check should be run again on a block with any user transactions in it to pass. But it shouldn't error.

sebastianst commented on op-chain-ops/cmd/check-jovian/main.go:
You could also check that the block header has the correct extra data format (length and encoded EIP1559 params & min base fee).

geoknee replied:
[fd70f96](https://github.com/ethereum-optimism/optimism/pull/18269/commits/fd70f960d5803a495e59f3874a3a2739ca901750)

geoknee committed:
op-chain-ops: add extra data format validation to check-jovian command

sebastianst commented on op-chain-ops/cmd/check-jovian/main.go:
```diff
@@ -149,6 +151,48 @@ func checkBlockHeader(ctx context.Context, env *actionEnv) error {
+	}
+
+	// Decode minimum base fee
+	minBaseFee := binary.BigEndian.Uint64(extra[9:17])
```
Why don't you just use any of `ValidateOptimismExtraData` or `DecodeOptimismExtraData` from op-geth package [eip1559](https://github.com/ethereum-optimism/op-geth/blob/0b2dd7a070f63a0c3a0941bd2000b9bb1110c3a6/consensus/misc/eip1559/eip1559_optimism.go?plain=1#L18-L44)?

geoknee replied:
[8cca10c](https://github.com/ethereum-optimism/optimism/pull/18269/commits/8cca10c1163e76618dd2361f2ac96f675fafc8ac)

geoknee committed:
op-chain-ops: enhance block header validation in check-jovian command to differentiate between zero and non-zero BlobGasUsed

geoknee committed:
refactor using op-geth library fns

sebastianst commented on op-chain-ops/cmd/check-jovian/main.go:
```diff
@@ -0,0 +1,209 @@
+
+	// A non-zero BlobGasUsed is hard evidence of Jovian being active
+	// A zero value is inconclusive (could be an empty block or pre-Jovian)
+	if *latest.BlobGasUsed == 0 {
```
We can make the check fail if the last transaction in the block is a non-deposit tx.

geoknee replied:
[5744d9d](https://github.com/ethereum-optimism/optimism/pull/18269/commits/5744d9df032c8918e45a53f13aee1902f4b4d5ba) (it's only an error if there are user txs, or more than one tx in the block).

geoknee committed:
Rename checkBlockHeader to checkBlock and improve Jovian activation
checks

geoknee committed:
Add secret-key flag to send tx-to-self

When a hex secret key is supplied, send a signed tx-to-self on L2, wait
up to 2 minutes for it to be mined, and use its block for the
BlobGasUsed check. Normalize 0x prefix and surface errors on failure.

geoknee committed:
lint

geoknee committed:
Require non-zero BlobGasUsed for Jovian

sebastianst commented on op-chain-ops/cmd/check-jovian/main.go:
```diff
@@ -138,11 +157,87 @@ func checkL1Block(ctx context.Context, env *actionEnv) error {
+			return fmt.Errorf("failed to get network id: %w", err)
+		}
+
+		tx := types.NewTransaction(nonce, toAddr, value, gasLimit, gasPrice, nil)
```
this is creating a legacy transaction, let's better use a modern `DynamicFeeTx` with `types.NewTx(&types.DynamicFeeTx{...})`

geoknee replied:
Should be fixed by [56e27d9](https://github.com/ethereum-optimism/optimism/pull/18269/commits/56e27d955416f451a7c03efacada7c58dabd38f6)

sebastianst commented on op-chain-ops/cmd/check-jovian/main.go:
I don't love that we're always reimplementing simple tx sending code with ~100 LoCs for every check script again... Wondering if there's an opportunity to extract this into a function in a `cmd/common` sub-package.

geoknee replied:
Fair, I realised we can just construct a quick txmgr and use that [56e27d9](https://github.com/ethereum-optimism/optimism/pull/18269/commits/56e27d955416f451a7c03efacada7c58dabd38f6)

sebastianst replied:
even better

sebastianst commented on op-chain-ops/cmd/check-jovian/main.go:
```diff
@@ -138,11 +157,87 @@ func checkL1Block(ctx context.Context, env *actionEnv) error {
+
 	bgu := latest.BlobGasUsed()
 	if bgu == nil {
 		return fmt.Errorf("block %d has nil BlobGasUsed field", latest.Number)
```
Github doesn't let me comment on line 253: instead of just checking that the blobGasUsed is != 0 we could also calculate the expected DA footprint using `types.CalcDAFootprint`.

geoknee replied:
[56b9160](https://github.com/ethereum-optimism/optimism/pull/18269/commits/56b91603f385d9f20d01b80042332c78399371c2)

geoknee committed:
Use txmgr to send and wait for tx in check-jovian

Replace manual key parsing, signing, and bind.WaitMined with a
SimpleTxManager (txmgr). Add l2endpoint to env and wire txmgr config,
using txmgr.Send to submit and await the self-transfer receipt.

geoknee committed:
Validate BlobGasUsed against DA footprint

geoknee committed:
Use provided context for transaction send

sebastianst commented on op-chain-ops/cmd/check-jovian/main.go:
```diff
@@ -0,0 +1,291 @@
+			"blockNumber", latest.Number(),
+			"note", "Zero could indicate an empty block or pre-Jovian state")
+
+		if len(txs) > 1 && txs[len(txs)-1].Type() == types.DepositTxType {
```
```suggestion
		if len(txs) > 1 && txs[len(txs)-1].Type() != types.DepositTxType {
```
the last tx must _not_ be a deposit tx for it to be a user tx.

But note that you could change the outer branch on line 221 just be `if len(tx) == 1` and then only print the warning log, and let the `else` branch also take care of 0 DA footprint.

geoknee replied:
Good spot, I took your second suggestion.

sebastianst commented on op-chain-ops/cmd/check-jovian/main.go:
```diff
@@ -0,0 +1,291 @@
+
+		if len(txs) > 1 && txs[len(txs)-1].Type() == types.DepositTxType {
+			return fmt.Errorf("User transactions in block but header.BlobGasUsed was zero, impossible if Jovian is active.")
+		}
```
```suggestion
	if len(txs) == 1 {
		env.log.Warn("Block has no user txs - inconclusive for Jovian activation",
			"blockNumber", latest.Number(),
			"note", "Zero could indicate an empty block or pre-Jovian state")
```

geoknee committed:
Update op-chain-ops/cmd/check-jovian/main.go

Co-authored-by: Sebastian Stammler <seb@oplabs.co>

geoknee committed:
Remove comment about BlobGasUsed

geoknee committed:
Document secret key option for check-jovian

Add usage examples for CHECK_JOVIAN_SECRET env var and the --secret flag

geoknee committed:
Return error on zero DA scalar; harden blobGasUsed

Treat a DA footprint gas scalar of 0 in L1Block as an error (Jovian
should not allow it). Delay calling t.From() until after creating the tx
manager and check the BlobGasUsed pointer before dereferencing to avoid
nil derefs.

geoknee committed:
Annotate txmgr logger with component field

geoknee committed:
Return error for blocks with no transactions

Treat blocks with zero transactions as an error instead of proceeding.
Retain the inconclusive warning for single-transaction blocks. Compute
DA footprint only for blocks with multiple transactions.

geoknee committed:
Merge remote-tracking branch 'origin/develop' into gk/check-jovian
